### PR TITLE
Penv-56: the platform requester

### PR DESCRIPTION
**Pull request by Tsovak**
**- What I did**
Develop the new utility for send request to the platform 

**- How I did it**
created new cmd with tests
**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: application/cmd/requester/requster_test.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	_ = os.Remove(s.userKeyFile.Name())
+}
+
+func (s *requesterSuiteTest) TestRequester_HelpWorks() {
+	sout, _ := runCmd("--help")
+	assert.Contains(s.T(), sout, "Usage of")
+}
+
+func (s *requesterSuiteTest) TestRequester_AllArgsPassedSuccessfully() {
+
```
Comment (Tsovak) [inline application/cmd/requester/requster_test.go]: delete this line

File: application/cmd/requester/main.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	pflag.StringVarP(&memberKeysPath, "memberkeys", "k", "", "Path to member key paramsFile")
+	pflag.StringVarP(&apiURL, "url", "u", "", "API URL. for example http://localhost:19101/api/rpc")
+	pflag.StringVarP(&inputRequestParams, "request", "r", "", "The request body or path to request paramsFile")
+	pflag.BoolVarP(&shouldPastePrivateKey, "autocompleteseed", "s", false, "Should replace seed to correct value")
+	pflag.BoolVarP(&shouldPasteSeed, "autocompletekey", "p", false, "Should replace publicKey to correct value")
+	pflag.Parse()
+}
+
+func logFatal(args ...interface{}) {
+	log.Fatal(args)
```
Comment (SergeyRadist) [inline application/cmd/requester/main.go]: logFatal is not needed, better use panic instead
btw log.Fatal is Print() followed by os.Exit(1)

File: application/cmd/requester/howto.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
+# The requester
```
Comment (megge-dream) [inline application/cmd/requester/howto.md]: Files with description we usually name README.md

File: application/cmd/requester/utils.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+func runCmd(args ...string) (string, error) {
+	cmd := exec.Command(binLocation, args...)
+	out, err := cmd.CombinedOutput()
+	return string(out), err
+}
+
+func getArgs(key, url, requestParams string, shouldPrivateKey, shouldSeed bool) []string {
+	cmd := make([]string, 5)
+	cmd[0] = fmt.Sprintf("-k=%v", key)
```
Comment (LinnikD) [inline application/cmd/requester/utils.go]: why not use cobra, like other binaries?

File: application/cmd/requester/utils.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+	cmd[1] = fmt.Sprintf("-u=%v", url)
+	cmd[2] = fmt.Sprintf("-r=%v", requestParams)
+	cmd[3] = fmt.Sprintf("-p=%v", shouldPrivateKey)
+	cmd[4] = fmt.Sprintf("-s=%v", shouldSeed)
+	return cmd
+}
+
+func writePrivateKeyToFile(pathToFile string) {
```
Comment (LinnikD) [inline application/cmd/requester/utils.go]: I think this function already exists in the project

File: application/cmd/requester/requster_test.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+func (s *requesterSuiteTest) TestRequester_UserShouldPassRequestParams() {
+	sout, err := runCmd(getArgs(s.userKeyFile.Name(), s.ts.URL, "/path/to/nonexisting/requestparams", true, true)...)
+	assert.Error(s.T(), err)
+	assert.Contains(s.T(), sout, "Cannot unmarshal request")
+}
+
+func (s *requesterSuiteTest) TestRequester_UserCanPassRequestParamsJson() {
+	sout, err := runCmd(getArgs(s.userKeyFile.Name(), s.ts.URL, createMemberRequestExample, true, true)...)
+	assert.NoError(s.T(), err)
+	assert.Contains(s.T(), sout, "requestReference")
```
Comment (LinnikD) [inline application/cmd/requester/requster_test.go]: is there a case when we send a correct request - and test server receives it with seed and signature?

Comment (Tsovak) [reply application/cmd/requester/requster_test.go]: look at TestRequester_AllArgsPassedSuccessfully

File: application/cmd/requester/utils.go (hunk 5 above)
Comment (Tsovak) [reply application/cmd/requester/utils.go]: are you talking about `generateKeysPairEthereum` in `insolar/main.go` file?  
I cannot use that. I need to extract to another function and use in both of them or write it again.

File: application/cmd/requester/cmd/root.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+	shouldPasteSeed      bool
+	shouldPastePublicKey bool
+	verbose              bool
+	memberPrivateKey     crypto.PrivateKey
+	request              *requester.ContractRequest
+	rootCmd              *cobra.Command
+)
+
+func init() {
+	rootCmd = getRequesterCommand()
```
Comment (nordicdyno) [inline application/cmd/requester/cmd/root.go]: move this to Execute function

File: application/cmd/requester/utils_test.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+	"net/http"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/insolar/insolar/insolar/secrets"
+)
+
+const (
+	binLocation                string = "./../../../bin/requester"
```
Comment (nordicdyno) [inline application/cmd/requester/utils_test.go]: don't test binary, test code

File: application/cmd/requester/utils_test.go [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+	cmd := make([]string, 5)
+	cmd[0] = fmt.Sprintf("-k=%v", key)
+	cmd[1] = fmt.Sprintf("-u=%v", url)
+	cmd[2] = fmt.Sprintf("-r=%v", requestParams)
+	cmd[3] = fmt.Sprintf("-p=%v", shouldPrivateKey)
+	cmd[4] = fmt.Sprintf("-s=%v", shouldSeed)
+	return cmd
+}
+
+func writePrivateKeyToFile(pathToFile string) {
```
Comment (nordicdyno) [inline application/cmd/requester/utils_test.go]: put asserts to `testdata` dir, don't generate them there is no reason for it here

File: application/cmd/requester/cmd/root.go [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+	parseInputParams(rootCmd)
+}
+
+func parseInputParams(cmd *cobra.Command) {
+	cmd.Flags().StringVarP(&memberKeysPath, "memberkeys", "k", "", "Path to member key")
+	cmd.Flags().StringVarP(&apiURL, "url", "u", "", "API URL. for example http://localhost:19101/api/rpc")
+	cmd.Flags().StringVarP(&inputRequestParams, "request", "r", "", "The request body or path to request params file")
+	cmd.Flags().BoolVarP(&shouldPasteSeed, "autocompleteseed", "s", true, "Should replace seed to correct value")
```
[the remark this answers is not in the document]
Comment (Tsovak) [reply application/cmd/requester/cmd/root.go]: no, we decided to left it here.

File: application/cmd/requester/cmd/root.go [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+		inputRequestParams = string(fileContent)
+	}
+
+	err = json.Unmarshal([]byte(inputRequestParams), &request)
+	if err != nil {
+		log.Fatal("Cannot unmarshal request. ["+inputRequestParams+"]", err)
+	}
+}
+
+func isUrl(str string) (bool, error) {
```
[the remark this answers is not in the document]
Comment (Tsovak) [reply application/cmd/requester/cmd/root.go]: if you pass `http://localhost:75000` the url.ParseRequestURI won't find the problem.

File: application/cmd/requester/utils_test.go (hunk 8 above)
Comment (Tsovak) [reply application/cmd/requester/utils_test.go]: it's not able to read cobra output. I already tried this way. I decided to run cms.Execute() in tests. It works correctly but not able to catch the output

File: application/cmd/requester/main.go [hunk 12]
```diff
@@ -0,0 +1,27 @@
+// Copyright 2020 Insolar Network Ltd.
```
Comment (megge-dream) [inline application/cmd/requester/main.go]: We have new license now, update please

File: application/api/requester/requester.go [hunk 13]
```diff
@@ -17,17 +17,20 @@ package requester
 import (
 	"bytes"
 	"context"
+	"crypto"
```
Comment (megge-dream) [inline application/api/requester/requester.go]: Why replace `x-crypto` package to `crypto`?

File: application/cmd/requester/cmd/root.go [hunk 14]
```diff
@@ -0,0 +1,186 @@
+// Copyright 2020 Insolar Network Ltd.
```
Comment (megge-dream) [inline application/cmd/requester/cmd/root.go]: Why this file inside another directory `cmd` in `requester` package?

Comment (Tsovak) [reply application/cmd/requester/cmd/root.go]: cobra library needs this solution

File: application/cmd/requester/requster_test.go [hunk 15]
```diff
@@ -0,0 +1,95 @@
+//  Copyright 2020 Insolar Network Ltd.
+//  All rights reserved.
+//  This material is licensed under the Insolar License version 1.0,
+//  available at https://github.com/insolar/insolar/blob/master/network/LICENSE.md.
+
+// +build requestertest
```
Comment (nordicdyno) [inline application/cmd/requester/requster_test.go]: please avoid build tag

File: application/cmd/requester/requster_test.go [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+func (s *requesterSuiteTest) TestRequester_UserCanPassRequestParamsJson() {
+	sout, err := runCmd(getArgs(s.userKeyFile.Name(), s.ts.URL, createMemberRequestExample, true, true)...)
+	assert.NoError(s.T(), err)
+	assert.Contains(s.T(), sout, "requestReference")
+}
+
+func TestAllRequesterTests(t *testing.T) {
+	suite.Run(t, new(requesterSuiteTest))
```
Comment (nordicdyno) [inline application/cmd/requester/requster_test.go]: i don't like suits, at least they don't work very well with go toolset

File: application/cmd/requester/utils_test.go (hunk 9 above)
Comment (nordicdyno) [reply application/cmd/requester/utils_test.go]: I still think it's not a test job generate asserts in this case.

File: application/cmd/requester/utils_test.go (hunk 8 above)
Comment (nordicdyno) [reply application/cmd/requester/utils_test.go]: you can read cobra's output if set `SetOutput` on command (parent  one for example)

Comment (Tsovak) [reply application/cmd/requester/utils_test.go]: but I need a requester output also

File: application/cmd/requester/utils_test.go (hunk 9 above)
Comment (Tsovak) [reply application/cmd/requester/utils_test.go]: it's kind of hidden dependency antipattern. I really think that using the created files is a bad idea. 
I'll use if you want. No problem

File: application/cmd/requester/requster_test.go (hunk 16 above)
Comment (Tsovak) [reply application/cmd/requester/requster_test.go]: what do you prefer?  MainTest?

File: Makefile [hunk 17]
```diff
@@ -106,7 +107,6 @@ vendor: ## update vendor dependencies
 .PHONY: build
 build: $(BIN_DIR) $(INSOLARD) $(INSOLAR) $(INSGOCC) $(PULSARD) $(TESTPULSARD) $(INSGORUND) $(HEALTHCHECK) $(BENCHMARK) ## build all binaries
```
Comment (megge-dream) [inline Makefile]: Add `$(REQUESTER)` here

File: application/cmd/requester/capturer_test.go [hunk 18]
```diff
@@ -0,0 +1,78 @@
+//  Copyright 2020 Insolar Network Ltd.
+//  All rights reserved.
+//  This material is licensed under the Insolar License version 1.0,
+//  available at https://github.com/insolar/insolar/blob/master/master/LICENSE.md.
```
Comment (megge-dream) [inline application/cmd/requester/capturer_test.go]: Update to `// available at https://github.com/insolar/insolar/blob/master/LICENSE.md.`

File: application/cmd/requester/cmd/root.go [hunk 19]
```diff
@@ -0,0 +1,195 @@
+//  Copyright 2020 Insolar Network Ltd.
+//  All rights reserved.
+//  This material is licensed under the Insolar License version 1.0,
+//  available at https://github.com/insolar/insolar/blob/master/master/LICENSE.md.
```
Comment (megge-dream) [inline application/cmd/requester/cmd/root.go]: Update to `// available at https://github.com/insolar/insolar/blob/master/LICENSE.md.`

File: application/cmd/requester/main.go [hunk 20]
```diff
@@ -0,0 +1,18 @@
+//  Copyright 2020 Insolar Network Ltd.
+//  All rights reserved.
+//  This material is licensed under the Insolar License version 1.0,
+//  available at https://github.com/insolar/insolar/blob/master/master/LICENSE.md.
```
Comment (megge-dream) [inline application/cmd/requester/main.go]: Update to `// available at https://github.com/insolar/insolar/blob/master/LICENSE.md.`

File: application/cmd/requester/requster_test.go [hunk 21]
```diff
@@ -0,0 +1,95 @@
+//  Copyright 2020 Insolar Network Ltd.
+//  All rights reserved.
+//  This material is licensed under the Insolar License version 1.0,
+//  available at https://github.com/insolar/insolar/blob/master/master/LICENSE.md.
```
Comment (megge-dream) [inline application/cmd/requester/requster_test.go]: Update to `// available at https://github.com/insolar/insolar/blob/master/LICENSE.md.`

File: application/cmd/requester/utils_test.go [hunk 22]
```diff
@@ -0,0 +1,130 @@
+//  Copyright 2020 Insolar Network Ltd.
+//  All rights reserved.
+//  This material is licensed under the Insolar License version 1.0,
+//  available at https://github.com/insolar/insolar/blob/master/master/LICENSE.md.
```
Comment (megge-dream) [inline application/cmd/requester/utils_test.go]: Update to `// available at https://github.com/insolar/insolar/blob/master/LICENSE.md.`

File: application/cmd/requester/utils_test.go (hunk 9 above)
Comment (nordicdyno) [reply application/cmd/requester/utils_test.go]: I disagree. test asserts is well known and good practice, and not antipattern here

File: application/cmd/requester/requster_test.go (hunk 16 above)
Comment (nordicdyno) [reply application/cmd/requester/requster_test.go]: yes, I prefer standard (and better supported by tools)  MainTest, I will ok with suits if tooling will work with them properly (like ability to run one test separately)

File: application/cmd/requester/utils_test.go [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+			if _, err := fmt.Fprint(w, "{\"requestReference\" :\"sddasacs\"'}"); err != nil {
+				log.Println(err.Error())
+			}
+		}
+	})
+	return r
+}
+
+func runCmd(args ...string) (string, error) {
+	// cmd := exec.Command(binLocation, args...)
```
Comment (megge-dream) [inline application/cmd/requester/utils_test.go]: Why comment and not delete?

**Pull request merged**